### PR TITLE
sql/multiregion: Add basic validation for zone config extension setting

### LINKED ISSUE #85015: multiregion: add validation for zone config extension
We should validate the zone config extension set by the user won't affect certain existing multi-region primitives. 

As suggested in https://github.com/cockroachdb/cockroach/pull/83605#pullrequestreview-1044433593, we may want to add the following validation:

1. REGIONAL extensions don't set lease_preferences
2. REGIONAL IN extensions don't set lease_preferences that disagree with their home region. I.e. After the extension gets applied, the home region is unchanged.
3. REGIONAL IN extensions don't set replica constraints or voter contains that disagree with their home region.
4. Extensions can't break a database's survivability goal.

If we find out the user gives a zone config extension that violate the rules above, we return an error and reject the incoherent extension. We want to guide the user towards a configuration that makes sense.

Jira issue: CRDB-18004
Epic CRDB-14616

### PR DESCRIPTION
**Pull request by ZhouXing19**
We added constraints on setting the zone config extension so that it won't break
the original region configs (such as home region and survival goal requirement).

We didn't exhaust all rules but added the following:

1. REGIONAL extensions don't set lease_preferences
2. REGIONAL IN extensions don't set lease_preferences that disagree with their home
region. I.e. After the extension gets applied, the home region is unchanged.
3. REGIONAL IN extensions don't set replica constraints or voter contains that disagree
with their home region.
4. Extensions can't set the number of replicas and voters lower than the ones
required by the database's survival goal.

Note that the validation for zone config extension is different from other
zone config settings in that it will influence the derived zone config for not
just the database, but also the tables and partitions under it. So we need to
ensure the validation pass for all these schemas before truly write the zone
config updates.

fixes #85015

Release justification: important validation for a multiregion funtionality.
Release note: add basic validation for zone config extension setting

### CODE REVIEW

**Review by rafiss — COMMENTED** [review]
just wondering, should some of this validation go into `validateMultiRegion` in pkg/sql/catalog/dbdesc/database_desc.go. this is validation that's lower-level at the descriptor level, so it prevents any other change to the descriptor from doing something invalid.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86538#-:-NAe-Lkk5EdGJ1pyNOh-:b-jx4fiy)* status: :shipit: complete! 0 of 0 LGTMs obtained

**Review by rafiss — COMMENTED** [review]
i think we might want the validation in both places. the ones in this PR should show user-friendly errors, and the ones in `validateMultiRegion` should show assertion errors.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86538#-:-NAe0q_tCxXgAXvZPjM4:b-jc2spq)* status: :shipit: complete! 0 of 0 LGTMs obtained

**Review by ZhouXing19 — COMMENTED** [review]
I'm not sure if it's necessary to add the validation in the descriptor level, as the validation we have here are mostly for the zone config itself. The logic here is similar to the existing validations for the `set zone config` logic, which IIUC are mostly done in the `sql` package rather than the descriptor level ([example](https://github.com/cockroachdb/cockroach/blob/448352ce4ed71a58e7701e82c786dd5152498310/pkg/sql/set_zone_config.go#L1045)).  

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86538#-:-NAjyraZ4Ky0NIm9eqwU:b-ucwwdg)* status: :shipit: complete! 0 of 0 LGTMs obtained

**Comment by ZhouXing19** [comment]
For the requirement that `Extensions can't break a database's survivability goal`, I'm not sure how to ensure that, so I translate it into `Extensions can't set the number of replicas and voters lower than the ones
required by the database's survival goal.` Not sure if that makes sense or is too limited.

**Comment by rafiss** [comment]
this lgtm, but i could also use an opinion from @RichardJCai

File: pkg/sql/alter_database.go [hunk 1]
```diff
@@ -2196,22 +2196,38 @@ func (n *alterDatabaseSetZoneConfigExtensionNode) startExec(params runParams) er
...
 
-	// Update the database's zone configuration.
-	if err := ApplyZoneConfigFromDatabaseRegionConfig(
+	// We need to validate that the zone config extension won't violate the original
+	// region config before we truly write the zone config the update. Since the
+	// zone config extension can affect the final zone config of the database AND
+	// tables in it, we now follow these steps:
+	// 1. Generate and Validate the newly created zone config for the database.
+	// 2. For each table under this database, validate the newly derived zone
+	// config. After the validation passed for all table, we are now sure that the
```
Comment (RichardJCai) [inline pkg/sql/alter_database.go:2205]: nit: all tables

File: pkg/sql/catalog/multiregion/region_config.go [hunk 2]
```diff
@@ -153,28 +153,79 @@ func (r *RegionConfig) ZoneConfigExtensions() descpb.ZoneConfigExtensions {
...
 		zc = extendZoneCfg(zc, ext)
 	}
-	return zc
+
+	// TODO(janexing): to ensure that the zone config extension won't break the
+	// survival goal, here we only add restriction on the number of voters and
+	// replicas. We may want to consider adding constraint to their distribution
+	// across zones/regions as well.
+	if zc.NumVoters != nil && *zc.NumVoters < numVoters {
+		return zonepb.ZoneConfig{}, errors.Newf("zone config extension "+
```
Comment (RichardJCai) [inline pkg/sql/catalog/multiregion/region_config.go:219]: nit: cannot set num_voters lower than ...

File: pkg/sql/catalog/multiregion/region_config.go [hunk 3]
```diff
@@ -153,28 +153,79 @@ func (r *RegionConfig) ZoneConfigExtensions() descpb.ZoneConfigExtensions {
...
+			"cannot set a voter number %v that is lower than the one required for the "+
+			"survival goal: %v with goal %v\n", *zc.NumVoters, numVoters, r.SurvivalGoal())
+	}
+	if zc.NumReplicas != nil && *zc.NumReplicas < numReplicas {
+		return zonepb.ZoneConfig{}, errors.Newf("zone config extension "+
+			"cannot set a replica number %v that is lower than the one required for the "+
```
Comment (RichardJCai) [inline pkg/sql/catalog/multiregion/region_config.go:225]: num_replicas just to be explicit about the setting name

File: pkg/sql/catalog/multiregion/region_config.go [hunk 4]
```diff
@@ -468,3 +519,109 @@ func CanDropRegion(name catpb.RegionName, config RegionConfig) error {
...
+
+	// If the extension does not inherit that field from the original zone config,
+	// we will use the one in the extension, and hence we need to validate that
+	// field.
+	// Lease preferences.
+	if !(ext.ShouldInheritLeasePreferences(&zc)) && len(zc.LeasePreferences) > 0 && len(ext.LeasePreferences) > 0 {
+		oriHomeRegionLeasePreference, idx1 := getHomeRegionLeasePreference(zc.LeasePreferences, homeRegion)
+		extHomeRegionLeasePreference, idx2 := getHomeRegionLeasePreference(ext.LeasePreferences, homeRegion)
+		if idx1 != -1 {
+			// If can' find a lease preference for the home region in the zone config
```
Comment (RichardJCai) [inline pkg/sql/catalog/multiregion/region_config.go:575]: nit: if we can't

File: pkg/sql/catalog/multiregion/region_config.go [hunk 5]
```diff
@@ -468,3 +519,109 @@ func CanDropRegion(name catpb.RegionName, config RegionConfig) error {
...
+	if homeRegion == "" {
+		return nil
+	}
+
+	// If the extension does not inherit that field from the original zone config,
+	// we will use the one in the extension, and hence we need to validate that
+	// field.
+	// Lease preferences.
+	if !(ext.ShouldInheritLeasePreferences(&zc)) && len(zc.LeasePreferences) > 0 && len(ext.LeasePreferences) > 0 {
+		oriHomeRegionLeasePreference, idx1 := getHomeRegionLeasePreference(zc.LeasePreferences, homeRegion)
```
Comment (RichardJCai) [inline pkg/sql/catalog/multiregion/region_config.go:572]: Rather we just use `originalHomeRegionLeasePreference` took me a second to understand what `ori` meant

File: pkg/sql/catalog/multiregion/region_config.go [hunk 6]
```diff
@@ -468,3 +519,109 @@ func CanDropRegion(name catpb.RegionName, config RegionConfig) error {
...
+					" region (%v) lease preference: %v\n", extHomeRegionLeasePreference, homeRegion, oriHomeRegionLeasePreference)
+			}
+		}
+	}
+
+	// Replica constraints.
+	if !(ext.ShouldInheritConstraints(&zc)) && len(zc.Constraints) > 0 && len(ext.Constraints) > 0 {
+		oriHomeRegionReplicaConstraint, ok1 := getHomeRegionConstraintConjunction(zc.Constraints, homeRegion)
+		extHomeRegionReplicaConstraint, ok2 := getHomeRegionConstraintConjunction(ext.Constraints, homeRegion)
+		if ok1 {
```
Comment (RichardJCai) [inline pkg/sql/catalog/multiregion/region_config.go:596]: This is a little bit hard to read.

Can we do `if ok1 && !ok2 { ... }` and split out the logic below
Also probably better to do name `ok1` as `homeRegionConstraintExistsForOldZC` and `ok2` as `homeRegionConstraintExistsForExt` as something like, I prefer verbosity here to make it clear.

**Review by RichardJCai — COMMENTED** [review]
Mostly LGTM good checks. Just some nits

File: pkg/sql/catalog/multiregion/region_config.go (hunk 6 above)
Comment (ZhouXing19) [reply pkg/sql/catalog/multiregion/region_config.go]: I changed it to return the reference of the constraint, so the second return value is not very much needed.

File: pkg/sql/alter_database.go (hunk 1 above)
Comment (ZhouXing19) [reply pkg/sql/alter_database.go]: Done.

File: pkg/sql/catalog/multiregion/region_config.go (hunk 3 above)
Comment (ZhouXing19) [reply pkg/sql/catalog/multiregion/region_config.go]: Done.

File: pkg/sql/catalog/multiregion/region_config.go (hunk 4 above)
Comment (ZhouXing19) [reply pkg/sql/catalog/multiregion/region_config.go]: Done.

File: pkg/sql/catalog/multiregion/region_config.go (hunk 5 above)
Comment (ZhouXing19) [reply pkg/sql/catalog/multiregion/region_config.go]: Done.

File: pkg/sql/catalog/multiregion/region_config.go (hunk 2 above)
Comment (ZhouXing19) [reply pkg/sql/catalog/multiregion/region_config.go:219]: Done.

**Review by RichardJCai — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86538)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten, @rafiss, @RichardJCai, and @ZhouXing19)

___
*[`pkg/sql/catalog/multiregion/region_config.go` line 527 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86538#-NAwcRx82AIpSAhGOzuI:-NAwcRx97KzXZkJNWeMu:bh37rqb) ([raw file](https://github.com/cockroachdb/cockroach/blob/70d0ac7d6334fffa4400368e897ab98a6dbbb8f0/pkg/sql/catalog/multiregion/region_config.go#L527)):*
> ```Go
> // If we found the constraint for the region, we return the corresponding
> // ConstraintsConjunction and "true" as the second boolean returned value.
> // If not found in the list, it returns an empty lease preference and "false".
> ```

This comment is out of date now

**Review by ZhouXing19 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86538)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten, @rafiss, and @RichardJCai)

___
*[`pkg/sql/catalog/multiregion/region_config.go` line 527 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/86538#-NAwcRx82AIpSAhGOzuI:-NBEf4CxCiem-wJVqc3Q:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/70d0ac7d6334fffa4400368e897ab98a6dbbb8f0/pkg/sql/catalog/multiregion/region_config.go#L527)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

This comment is out of date now
</blockquote></details>

Done.

**Comment by ZhouXing19** [comment]
TFTR!
bors r=RichardJCai

**Pull request merged**